In [145]:
from io import StringIO
import requests
import json
import pandas as pd
import re
from collections import defaultdict

%matplotlib inline

In [2]:
# The code was removed by DSX for sharing.

Unnamed: 0          CASE_STATUS  \
0           1  CERTIFIED-WITHDRAWN   
1           2  CERTIFIED-WITHDRAWN   
2           3  CERTIFIED-WITHDRAWN   
3           4  CERTIFIED-WITHDRAWN   
4           5            WITHDRAWN   

                                       EMPLOYER_NAME  \
0                             UNIVERSITY OF MICHIGAN   
1                             GOODMAN NETWORKS, INC.   
2                          PORTS AMERICA GROUP, INC.   
3  GATES CORPORATION, A WHOLLY-OWNED SUBSIDIARY O...   
4                          PEABODY INVESTMENTS CORP.   

                        SOC_NAME                     JOB_TITLE  \
0  BIOCHEMISTS AND BIOPHYSICISTS  POSTDOCTORAL RESEARCH FELLOW   
1               CHIEF EXECUTIVES       CHIEF OPERATING OFFICER   
2               CHIEF EXECUTIVES         CHIEF PROCESS OFFICER   
3               CHIEF EXECUTIVES   REGIONAL PRESIDEN, AMERICAS   
4               CHIEF EXECUTIVES  PRESIDENT MONGOLIA AND INDIA   

  FULL_TIME_POSITION  PREVAILING_WAGE    YEAR                 WORKSITE  \
0                  N          36067.0  2016.0      ANN ARBOR, MICHIGAN   
1                  Y         242674.0  2016.0             PLANO, TEXAS   
2                  Y         193066.0  2016.0  JERSEY CITY, NEW JERSEY   
3                  Y         220314.0  2016.0         DENVER, COLORADO   
4                  Y         157518.4  2016.0      ST. LOUIS, MISSOURI   

          lon        lat  
0  -83.743038  42.280826  
1  -96.698886  33.019843  
2  -74.077642  40.728158  
3 -104.990251  39.739236  
4  -90.199404  38.627003

In [139]:
df_data_2.head(10)

Unnamed: 0          CASE_STATUS  \
0           1  CERTIFIED-WITHDRAWN   
1           2  CERTIFIED-WITHDRAWN   
2           3  CERTIFIED-WITHDRAWN   
3           4  CERTIFIED-WITHDRAWN   
4           5            WITHDRAWN   
5           6  CERTIFIED-WITHDRAWN   
6           7  CERTIFIED-WITHDRAWN   
7           8  CERTIFIED-WITHDRAWN   
8           9  CERTIFIED-WITHDRAWN   
9          10            WITHDRAWN   

                                       EMPLOYER_NAME  \
0                             UNIVERSITY OF MICHIGAN   
1                             GOODMAN NETWORKS, INC.   
2                          PORTS AMERICA GROUP, INC.   
3  GATES CORPORATION, A WHOLLY-OWNED SUBSIDIARY O...   
4                          PEABODY INVESTMENTS CORP.   
5                            BURGER KING CORPORATION   
6                   BT AND MK ENERGY AND COMMODITIES   
7                    GLOBO MOBILE TECHNOLOGIES, INC.   
8                                 ESI COMPANIES INC.   
9                          LESSARD INTERNATIONAL LLC   

                        SOC_NAME  \
0  BIOCHEMISTS AND BIOPHYSICISTS   
1               CHIEF EXECUTIVES   
2               CHIEF EXECUTIVES   
3               CHIEF EXECUTIVES   
4               CHIEF EXECUTIVES   
5               CHIEF EXECUTIVES   
6               CHIEF EXECUTIVES   
7               CHIEF EXECUTIVES   
8               CHIEF EXECUTIVES   
9               CHIEF EXECUTIVES   

                                           JOB_TITLE FULL_TIME_POSITION  \
0                       POSTDOCTORAL RESEARCH FELLOW                  N   
1                            CHIEF OPERATING OFFICER                  Y   
2                              CHIEF PROCESS OFFICER                  Y   
3                        REGIONAL PRESIDEN, AMERICAS                  Y   
4                       PRESIDENT MONGOLIA AND INDIA                  Y   
5  EXECUTIVE V P, GLOBAL DEVELOPMENT AND PRESIDEN...                  Y   
6                            CHIEF OPERATING OFFICER                  Y   
7                           CHIEF OPERATIONS OFFICER                  Y   
8                                          PRESIDENT                  Y   
9                                          PRESIDENT                  Y   

   PREVAILING_WAGE    YEAR                 WORKSITE         lon        lat  
0          36067.0  2016.0      ANN ARBOR, MICHIGAN  -83.743038  42.280826  
1         242674.0  2016.0             PLANO, TEXAS  -96.698886  33.019843  
2         193066.0  2016.0  JERSEY CITY, NEW JERSEY  -74.077642  40.728158  
3         220314.0  2016.0         DENVER, COLORADO -104.990251  39.739236  
4         157518.4  2016.0      ST. LOUIS, MISSOURI  -90.199404  38.627003  
5         225000.0  2016.0           MIAMI, FLORIDA  -80.191790  25.761680  
6          91021.0  2016.0           HOUSTON, TEXAS  -95.369803  29.760427  
7         150000.0  2016.0     SAN JOSE, CALIFORNIA -121.886329  37.338208  
8         127546.0  2016.0           MEMPHIS, TEXAS         NaN        NaN  
9         154648.0  2016.0         VIENNA, VIRGINIA  -77.265260  38.901222

In [4]:
df_data_2[['CASE_STATUS', 'SOC_NAME']]
#use this to make a chart with two indexes

CASE_STATUS  \
0        CERTIFIED-WITHDRAWN   
1        CERTIFIED-WITHDRAWN   
2        CERTIFIED-WITHDRAWN   
3        CERTIFIED-WITHDRAWN   
4                  WITHDRAWN   
5        CERTIFIED-WITHDRAWN   
6        CERTIFIED-WITHDRAWN   
7        CERTIFIED-WITHDRAWN   
8        CERTIFIED-WITHDRAWN   
9                  WITHDRAWN   
10       CERTIFIED-WITHDRAWN   
11       CERTIFIED-WITHDRAWN   
12       CERTIFIED-WITHDRAWN   
13       CERTIFIED-WITHDRAWN   
14       CERTIFIED-WITHDRAWN   
15       CERTIFIED-WITHDRAWN   
16       CERTIFIED-WITHDRAWN   
17       CERTIFIED-WITHDRAWN   
18                 CERTIFIED   
19                 CERTIFIED   
20       CERTIFIED-WITHDRAWN   
21       CERTIFIED-WITHDRAWN   
22                 CERTIFIED   
23                 CERTIFIED   
24       CERTIFIED-WITHDRAWN   
25                 CERTIFIED   
26       CERTIFIED-WITHDRAWN   
27                 CERTIFIED   
28                 CERTIFIED   
29                 CERTIFIED   
...                      ...   
3002428            WITHDRAWN   
3002429            WITHDRAWN   
3002430            WITHDRAWN   
3002431            WITHDRAWN   
3002432            WITHDRAWN   
3002433            WITHDRAWN   
3002434            WITHDRAWN   
3002435            WITHDRAWN   
3002436            WITHDRAWN   
3002437            WITHDRAWN   
3002438            WITHDRAWN   
3002439            WITHDRAWN   
3002440            WITHDRAWN   
3002441            WITHDRAWN   
3002442            WITHDRAWN   
3002443            WITHDRAWN   
3002444            WITHDRAWN   
3002445                  NaN   
3002446                  NaN   
3002447                  NaN   
3002448                  NaN   
3002449                  NaN   
3002450                  NaN   
3002451                  NaN   
3002452                  NaN   
3002453                  NaN   
3002454                  NaN   
3002455                  NaN   
3002456                  NaN   
3002457                  NaN   

                                                  SOC_NAME  
0                            BIOCHEMISTS AND BIOPHYSICISTS  
1                                         CHIEF EXECUTIVES  
2                                         CHIEF EXECUTIVES  
3                                         CHIEF EXECUTIVES  
4                                         CHIEF EXECUTIVES  
5                                         CHIEF EXECUTIVES  
6                                         CHIEF EXECUTIVES  
7                                         CHIEF EXECUTIVES  
8                                         CHIEF EXECUTIVES  
9                                         CHIEF EXECUTIVES  
10                                        CHIEF EXECUTIVES  
11                                        CHIEF EXECUTIVES  
12                                        CHIEF EXECUTIVES  
13                                        CHIEF EXECUTIVES  
14                                        CHIEF EXECUTIVES  
15                                        CHIEF EXECUTIVES  
16                                      FINANCIAL MANAGERS  
17                                        CHIEF EXECUTIVES  
18                                        CHIEF EXECUTIVES  
19                                        CHIEF EXECUTIVES  
20                                        CHIEF EXECUTIVES  
21                                        CHIEF EXECUTIVES  
22                                        CHIEF EXECUTIVES  
23                                        CHIEF EXECUTIVES  
24                                        CHIEF EXECUTIVES  
25                                        CHIEF EXECUTIVES  
26                                        CHIEF EXECUTIVES  
27                                        CHIEF EXECUTIVES  
28                                        CHIEF EXECUTIVES  
29                                        CHIEF EXECUTIVES  
...                                                    ...  
3002428                                Physical Therapists  
3002429            

In [5]:
df_data_2['SOC_NAME'].value_counts()

Computer Systems Analysts                                                 291170
Computer Programmers                                                      226574
SOFTWARE DEVELOPERS, APPLICATIONS                                         221783
COMPUTER SYSTEMS ANALYSTS                                                 215353
Software Developers, Applications                                         192933
COMPUTER PROGRAMMERS                                                      171972
COMPUTER OCCUPATIONS, ALL OTHER                                           108555
Computer Occupations, All Other                                            69315
Software Developers, Systems Software                                      44500
SOFTWARE DEVELOPERS, SYSTEMS SOFTWARE                                      39356
Management Analysts                                                        38756
Financial Analysts                                                         34141
Computer Software Engineers,

In [6]:
df_data_2['SOC_NAME'].value_counts().count()

2132

In [66]:
cleandata1=df_data_2.copy()

In [67]:
cleandata1['SOC_NAME']=df_data_2['SOC_NAME'].str.lower()
cleandata1['EMPLOYER_NAME']=df_data_2['EMPLOYER_NAME'].str.lower()
cleandata1['JOB_TITLE']=df_data_2['JOB_TITLE'].str.lower()

Here, we transform some strings to lowercase. This is because there are duplicate entries in the dataset which in both upper and lower.
This increases redundancy

In [68]:
cleandata1['SOC_NAME'].value_counts()

computer systems analysts                                       506523
software developers, applications                               414716
computer programmers                                            398546
computer occupations, all other                                 177870
software developers, systems software                            83856
management analysts                                              66873
financial analysts                                               53540
accountants and auditors                                         52822
mechanical engineers                                             44236
network and computer systems administrators                      40087
database administrators                                          38982
market research analysts and marketing specialists               37737
electronics engineers, except computer                           36574
operations research analysts                                     34260
electr

There is still alot of redundancy we can exploit. We can generalize these strings remove specalized strings into more general form.
For example 'software engineer, senior' needs to be reduced to 'software engineer'.
This also applies to the other columns with string attributes.

In [70]:
cleandata1['SOC_NAME'].value_counts().count()

1585

We have reduced the number of names down to 1585

In [71]:
reducedf = pd.DataFrame({'SOC_NAME': cleandata1['SOC_NAME'].value_counts().index, 'Count':cleandata1['SOC_NAME'].value_counts().values})
#df['Counts'] = df.groupby(['SOC_NAME'])['Count'].transform('count') #I don't remember what I was trying to do here.
#df = df.set_index(['SOC_NAME'])
reducedf

Count                                           SOC_NAME
0     506523                          computer systems analysts
1     414716                  software developers, applications
2     398546                               computer programmers
3     177870                    computer occupations, all other
4      83856              software developers, systems software
5      66873                                management analysts
6      53540                                 financial analysts
7      52822                           accountants and auditors
8      44236                               mechanical engineers
9      40087        network and computer systems administrators
10     38982                            database administrators
11     37737  market research analysts and marketing special...
12     36574             electronics engineers, except computer
13     34260                       operations research analysts
14     34108                               electrical engineers
15     33526                 physicians and surgeons, all other
16     33387          computer software engineers, applications
17     27536          computer and information systems managers
18     26254                   computer occupations, all other*
19     26159         medical scientists, except epidemiologists
20     21994                                physical therapists
21     21245                      biochemists and biophysicists
22     19370                               industrial engineers
23     17426                           computer systems analyst
24     17101                                      statisticians
25     16367                   biological scientists, all other
26     16310                                 marketing managers
27     15970                                    civil engineers
28     15000                                     web developers
29     13367                                internists, general
...      ...                                                ...
1555       1                              helpers--electricians
1556       1                               industrial designers
1557       1                                  logistics manager
1558       1             electronics engineers, expert computer
1559       1                                 hardward engineers
1560       1                           soil and plant scientist
1561       1             mechanical engineers, r&d (acwia only)
1562       1             timing device assemblers and adjusters
1563       1   telecommunications line installers and repairers
1564       1                        biological science teachers
1565       1                                            dentist
1566       1              computer systems engineers/arquitects
1567       1                         secondary school teachers,
1568       1                           computer systems anaysts
1569       1                            counseling psychologist
1570       1                          software engineer, senior
1571       1  health diagnosing & treating practitioners, al...
1572       1                          set and exhibit engineers
1573       1            helpers, construction trades, all other
1574       1                            datebase administrators
1575       1               commercial and insdistrial designers
1576       1                 software developers, appllications
1577       1                     computer occupations,all other
1578       1  foreign language and literature teachers, post...
1579       1                                           dentists
1580       1                railroad conductors and yardmasters
1581       1  computer occupations, all other: information t...
1582       1                         computer network architect
1583       1  criminal justice and law enforcement teachers,...
1584       1          medical scientists, except epidemiologist

[1585 rows x 2 columns]

In [72]:
reducedf['Name1'] = ''

In [73]:
reducedf

Count                                           SOC_NAME Name1
0     506523                          computer systems analysts      
1     414716                  software developers, applications      
2     398546                               computer programmers      
3     177870                    computer occupations, all other      
4      83856              software developers, systems software      
5      66873                                management analysts      
6      53540                                 financial analysts      
7      52822                           accountants and auditors      
8      44236                               mechanical engineers      
9      40087        network and computer systems administrators      
10     38982                            database administrators      
11     37737  market research analysts and marketing special...      
12     36574             electronics engineers, except computer      
13     34260                       operations research analysts      
14     34108                               electrical engineers      
15     33526                 physicians and surgeons, all other      
16     33387          computer software engineers, applications      
17     27536          computer and information systems managers      
18     26254                   computer occupations, all other*      
19     26159         medical scientists, except epidemiologists      
20     21994                                physical therapists      
21     21245                      biochemists and biophysicists      
22     19370                               industrial engineers      
23     17426                           computer systems analyst      
24     17101                                      statisticians      
25     16367                   biological scientists, all other      
26     16310                                 marketing managers      
27     15970                                    civil engineers      
28     15000                                     web developers      
29     13367                                internists, general      
...      ...                                                ...   ...
1555       1                              helpers--electricians      
1556       1                               industrial designers      
1557       1                                  logistics manager      
1558       1             electronics engineers, expert computer      
1559       1                                 hardward engineers      
1560       1                           soil and plant scientist      
1561       1             mechanical engineers, r&d (acwia only)      
1562       1             timing device assemblers and adjusters      
1563       1   telecommunications line installers and repairers      
1564       1                        biological science teachers      
1565       1                                            dentist      
1566       1              computer systems engineers/arquitects      
1567       1                         secondary school teachers,      
1568       1                           computer systems anaysts      
1569       1                            counseling psychologist      
1570       1                          software engineer, senior      
1571       1  health diagnosing & treating practitioners, al...      
1572       1                          set and exhibit engineers      
1573       1            helpers, construction trades, all other      
1574       1                            datebase administrators      
1575       1               commercial and insdistrial designers      
1576       1                 software developers, appllications      
1577       1                     computer occupations,all other      
1578       1  foreign language and literature teachers, post...      
1579       1                                           dentists      
1580       1     

In [74]:
reducedf.iloc[3]['Count'] #example of accessing a location

177870

In [75]:
%%timeit
for index, row in reducedf.iterrows():
    names = row['SOC_NAME'].split(",")
    if(names[0].endswith('*')):
        reducedf.set_value([index],['Name1'],(names[0][:-1]))
    if not (names[0].endswith('s')):
        reducedf.set_value([index],['Name1'],(names[0]+'s'))
    else:
        reducedf.set_value([index],['Name1'],names[0])

1 loop, best of 3: 1.52 s per loop


In [76]:
reducedf

Count                                           SOC_NAME  \
0     506523                          computer systems analysts   
1     414716                  software developers, applications   
2     398546                               computer programmers   
3     177870                    computer occupations, all other   
4      83856              software developers, systems software   
5      66873                                management analysts   
6      53540                                 financial analysts   
7      52822                           accountants and auditors   
8      44236                               mechanical engineers   
9      40087        network and computer systems administrators   
10     38982                            database administrators   
11     37737  market research analysts and marketing special...   
12     36574             electronics engineers, except computer   
13     34260                       operations research analysts   
14     34108                               electrical engineers   
15     33526                 physicians and surgeons, all other   
16     33387          computer software engineers, applications   
17     27536          computer and information systems managers   
18     26254                   computer occupations, all other*   
19     26159         medical scientists, except epidemiologists   
20     21994                                physical therapists   
21     21245                      biochemists and biophysicists   
22     19370                               industrial engineers   
23     17426                           computer systems analyst   
24     17101                                      statisticians   
25     16367                   biological scientists, all other   
26     16310                                 marketing managers   
27     15970                                    civil engineers   
28     15000                                     web developers   
29     13367                                internists, general   
...      ...                                                ...   
1555       1                              helpers--electricians   
1556       1                               industrial designers   
1557       1                                  logistics manager   
1558       1             electronics engineers, expert computer   
1559       1                                 hardward engineers   
1560       1                           soil and plant scientist   
1561       1             mechanical engineers, r&d (acwia only)   
1562       1             timing device assemblers and adjusters   
1563       1   telecommunications line installers and repairers   
1564       1                        biological science teachers   
1565       1                                            dentist   
1566       1              computer systems engineers/arquitects   
1567       1                         secondary school teachers,   
1568       1                           computer systems anaysts   
1569       1                            counseling psychologist   
1570       1                          software engineer, senior   
1571       1  health diagnosing & treating practitioners, al...   
1572       1                          set and exhibit engineers   
1573       1            helpers, construction trades, all other   
1574       1                            datebase administrators   
1575       1               commercial and insdistrial designers   
1576       1                 software developers, appllications   
1577       1                     computer occupations,all other   
1578       1  foreign language and literature teachers, post...   
1579       1                                           dentists   
1580       1                railroad conductors and yardmasters   
1581       1  computer occupations, all other: information t...   
1582       1                         computer network 

In [79]:
cleandata1['SOC_NAME'].value_counts().count()

1585

In [98]:
(cleandata1.loc[(cleandata1['SOC_NAME']=='software developers, appllications')]) #an example of a query

Unnamed: 0 CASE_STATUS       EMPLOYER_NAME  \
894328      894329   CERTIFIED  meridiansoft, inc.   

                                  SOC_NAME                        JOB_TITLE  \
894328  software developers, appllications  software developer applications   

       FULL_TIME_POSITION  PREVAILING_WAGE    YEAR            WORKSITE  \
894328                  Y          62421.0  2015.0  LEWIS CENTER, OHIO   

              lon        lat  
894328 -83.010099  40.198388

This person messed up the SOC_NAME

In [80]:
reducedf['Name1'].value_counts()

software developers                                    37
secondary school teachers                              14
securities                                             14
sales representatives                                  13
electronics engineers                                  13
computer occupations                                   11
computer systems analysts                              10
computer software engineers                            10
elementary school teachers                             10
mechanical engineers                                    9
special education teachers                              8
electrical engineers                                    7
education administrators                                7
computer programmers                                    7
medical scientists                                      7
middle school teachers                                  7
cooks                                                   6
foreign langua

In [81]:
reducedf['Name1'].value_counts().count()

1134

We have now cut the number of names in half from the original number.

In [82]:
reducedf.sort_values(['Name1'])

Count                                           SOC_NAME  \
1286      1                                         13-2011.01   
1096      2                                            15-1121   
1243      1                                            15-1132   
1272      1   15-1199.01 sw quality assurance engnrs & testers   
1354      1                                         15-1199.01   
1393      1         15-1199.08, business intelligence analysts   
1314      1                                            17-2051   
1552      1                                            17-2072   
1461      1                                            27-3031   
1426      1         29-1064.00-obstetricians and gynecologists   
1319      1              <font><font>carpinteros</font></font>   
1071      2                                        able seamen   
1026      2                                         accountant   
74     3958                                        accountants   
7     52822                           accountants and auditors   
564      26                                             actors   
80     3486                                          actuaries   
1539      1                                     acupuncturists   
882       4     adhesive bonding machine operators and tenders   
821       6  administrative law judges, adjudicators, and h...   
152    1361                   administrative services managers   
323     230  adult basic and secondary education and litera...   
395     116   adult basic and secondary education and literacy   
1113      2  adult basic second educ and lit teach and instruc   
485      50  adult literacy, remedial education, and ged te...   
1322      1               advertising and promortions managers   
94     2925                advertising and promotions managers   
1126      2                 advertising and promotions manager   
342     195                           advertising sales agents   
351     179   aerospace engineering and operations technicians   
...     ...                                                ...   
1388      1                                         urologists   
1107      2        ushers, lobby attendants, and ticket takers   
901       4                                validation engineer   
422      90                               validation engineers   
113    2398                                      veterinarians   
599      19  veterinary assistants and laboratory animal ca...   
596      20        veterinary assistants and laboratory animal   
426      87           veterinary technologists and technicians   
1502      1               vice president of investor relations   
958       3                               video game designers   
1016      3       vocational education teachers, middle school   
626      16    vocational education teachers, secondary school   
331     215       vocational education teachers, postsecondary   
492      47        vocational education teachers postsecondary   
639      16                                    watch repairers   
878       4  water and wastewater treatment plant and syste...   
1008      3                         water/wastewater engineers   
437      81                                 web administrators   
1052      2                                      web developer   
28    15000                                     web developers   
1043      2       weighers, measurers, checkers, and samplers,   
638      16           welders, cutters, solderers, and brazers   
1204      1   welding, soldering, and brazing machine setters,   
1237      1  welding, soldering, and brazing machine setter...   
205     746  wholesale and retail buyers, except farm products   
1259      1                              wind energy engineers   
1296      1                   wind turbine service technicians   
708      10                             woodworkers, all other   
181     981                                wri

In [83]:
reducedf['Name2'] = ""

In [84]:
%%timeit
regex = re.compile('[^a-z\s]')

for index, row in reducedf.iterrows():
    reducedf.set_value([index],['Name2'],(regex.sub('', row['Name1'])))

1 loop, best of 3: 1.52 s per loop


In [85]:
reducedf.sort_values(['Name1'])

Count                                           SOC_NAME  \
1286      1                                         13-2011.01   
1096      2                                            15-1121   
1243      1                                            15-1132   
1272      1   15-1199.01 sw quality assurance engnrs & testers   
1354      1                                         15-1199.01   
1393      1         15-1199.08, business intelligence analysts   
1314      1                                            17-2051   
1552      1                                            17-2072   
1461      1                                            27-3031   
1426      1         29-1064.00-obstetricians and gynecologists   
1319      1              <font><font>carpinteros</font></font>   
1071      2                                        able seamen   
1026      2                                         accountant   
74     3958                                        accountants   
7     52822                           accountants and auditors   
564      26                                             actors   
80     3486                                          actuaries   
1539      1                                     acupuncturists   
882       4     adhesive bonding machine operators and tenders   
821       6  administrative law judges, adjudicators, and h...   
152    1361                   administrative services managers   
323     230  adult basic and secondary education and litera...   
395     116   adult basic and secondary education and literacy   
1113      2  adult basic second educ and lit teach and instruc   
485      50  adult literacy, remedial education, and ged te...   
1322      1               advertising and promortions managers   
94     2925                advertising and promotions managers   
1126      2                 advertising and promotions manager   
342     195                           advertising sales agents   
351     179   aerospace engineering and operations technicians   
...     ...                                                ...   
1388      1                                         urologists   
1107      2        ushers, lobby attendants, and ticket takers   
901       4                                validation engineer   
422      90                               validation engineers   
113    2398                                      veterinarians   
599      19  veterinary assistants and laboratory animal ca...   
596      20        veterinary assistants and laboratory animal   
426      87           veterinary technologists and technicians   
1502      1               vice president of investor relations   
958       3                               video game designers   
1016      3       vocational education teachers, middle school   
626      16    vocational education teachers, secondary school   
331     215       vocational education teachers, postsecondary   
492      47        vocational education teachers postsecondary   
639      16                                    watch repairers   
878       4  water and wastewater treatment plant and syste...   
1008      3                         water/wastewater engineers   
437      81                                 web administrators   
1052      2                                      web developer   
28    15000                                     web developers   
1043      2       weighers, measurers, checkers, and samplers,   
638      16           welders, cutters, solderers, and brazers   
1204      1   welding, soldering, and brazing machine setters,   
1237      1  welding, soldering, and brazing machine setter...   
205     746  wholesale and retail buyers, except farm products   
1259      1                              wind energy engineers   
1296      1                   wind turbine service technicians   
708      10                             woodworkers, all other   
181     981                                wri

In [86]:
reducedf['Name2'].value_counts().count()

1118

Striping charactors did not help much.

At this point investigated a spellchecker in python. Was not able to get something to work within enviroment.

I am now going to consider removing entries that are unique, with a count of one.

In [87]:
dfName2Check = pd.DataFrame({'Name2': reducedf['Name2'].value_counts().index, 'Count':reducedf['Name2'].value_counts().values})

In [88]:
dfName2Check

Count                                              Name2
0        37                                software developers
1        14                                         securities
2        14                          secondary school teachers
3        13                              sales representatives
4        13                              electronics engineers
5        11                               computer occupations
6        10                        computer software engineers
7        10                          computer systems analysts
8        10                         elementary school teachers
9         9                               mechanical engineers
10        8                                                  s
11        8                         special education teachers
12        7                               electrical engineers
13        7                             middle school teachers
14        7                               computer programmers
15        7                                 medical scientists
16        7                           education administrators
17        6                                              cooks
18        6           foreign language and literature teachers
19        6                                    civil engineers
20        5                                       atmospherics
21        5                                           dentists
22        5                        computer network architects
23        5                        computer hardware engineers
24        4                    business operations specialists
25        4     healthcare practitioners and technical workers
26        4                            physicians and surgeons
27        4          computer and information systems managers
28        4                                      compensations
29        4                              biological scientists
...     ...                                                ...
1088      1                                         morticians
1089      1                           physical therapist aides
1090      1                                electrical enginers
1091      1              detectives and criminal investigators
1092      1                     environmental science teachers
1093      1                                  psychiatric aides
1094      1                              marketing specialists
1095      1                                medical seicentists
1096      1                               landscape architects
1097      1                                            singers
1098      1                          biochemist  biophysicists
1099      1                                 robotics engineers
1100      1            musical instrument repairers and tuners
1101      1           firstline supervisorsmanagers of helpers
1102      1                        operation research analysts
1103      1                    respiratory therapy technicians
1104      1                               skincare specialists
1105      1       cardiovascular technologists and technicians
1106      1                               copmuter programmers
1107      1  firstline supervisors of personal service workers
1108      1                                  nuclear engineers
1109      1                                          clinicals
1110      1                                        commodities
1111      1                                   project managers
1112      1                          healthcare social workers
1113      1                                         urologists
1114      1                                      orthodontists
1115      1                                        stonemasons
1116      1                            tile and marble setters
1117      1                                comuter occupations

[1118 rows x 2 columns]

Theres a problem, I am not getting a real representation of the occurence of names in the data. I now need to do something where I can get the actual number of occurences.

In [89]:
cleandata1

Unnamed: 0          CASE_STATUS  \
0                 1  CERTIFIED-WITHDRAWN   
1                 2  CERTIFIED-WITHDRAWN   
2                 3  CERTIFIED-WITHDRAWN   
3                 4  CERTIFIED-WITHDRAWN   
4                 5            WITHDRAWN   
5                 6  CERTIFIED-WITHDRAWN   
6                 7  CERTIFIED-WITHDRAWN   
7                 8  CERTIFIED-WITHDRAWN   
8                 9  CERTIFIED-WITHDRAWN   
9                10            WITHDRAWN   
10               11  CERTIFIED-WITHDRAWN   
11               12  CERTIFIED-WITHDRAWN   
12               13  CERTIFIED-WITHDRAWN   
13               14  CERTIFIED-WITHDRAWN   
14               15  CERTIFIED-WITHDRAWN   
15               16  CERTIFIED-WITHDRAWN   
16               17  CERTIFIED-WITHDRAWN   
17               18  CERTIFIED-WITHDRAWN   
18               19            CERTIFIED   
19               20            CERTIFIED   
20               21  CERTIFIED-WITHDRAWN   
21               22  CERTIFIED-WITHDRAWN   
22               23            CERTIFIED   
23               24            CERTIFIED   
24               25  CERTIFIED-WITHDRAWN   
25               26            CERTIFIED   
26               27  CERTIFIED-WITHDRAWN   
27               28            CERTIFIED   
28               29            CERTIFIED   
29               30            CERTIFIED   
...             ...                  ...   
3002428     3002429            WITHDRAWN   
3002429     3002430            WITHDRAWN   
3002430     3002431            WITHDRAWN   
3002431     3002432            WITHDRAWN   
3002432     3002433            WITHDRAWN   
3002433     3002434            WITHDRAWN   
3002434     3002435            WITHDRAWN   
3002435     3002436            WITHDRAWN   
3002436     3002437            WITHDRAWN   
3002437     3002438            WITHDRAWN   
3002438     3002439            WITHDRAWN   
3002439     3002440            WITHDRAWN   
3002440     3002441            WITHDRAWN   
3002441     3002442            WITHDRAWN   
3002442     3002443            WITHDRAWN   
3002443     3002444            WITHDRAWN   
3002444     3002445            WITHDRAWN   
3002445     3002446                  NaN   
3002446     3002447                  NaN   
3002447     3002448                  NaN   
3002448     3002449                  NaN   
3002449     3002450                  NaN   
3002450     3002451                  NaN   
3002451     3002452                  NaN   
3002452     3002453                  NaN   
3002453     3002454                  NaN   
3002454     3002455                  NaN   
3002455     3002456                  NaN   
3002456     3002457                  NaN   
3002457     3002458                  NaN   

                                             EMPLOYER_NAME  \
0                                   university of michigan   
1                                   goodman networks, inc.   
2                                ports america group, inc.   
3        gates corporation, a wholly-owned subsidiary o...   
4                                peabody investments corp.   
5                                  burger king corporation   
6                         bt and mk energy and commodities   
7                          globo mobile technologies, inc.   
8                                       esi companies inc.   
9                                lessard international llc   
10                                      h.j. heinz company   
11                                 dow corning corporation   
12                                        acushnet company   
13                                           biocair, inc.   
14                              newmont mining corporation   
15                                            vricon, inc.   
16                             cardiac science corporation   
17                                   westfield corporation   
18                                          quicklogix llc   
19                                   mcc

Going to go back to this df and create a copy. Will then over write the Soc_Name with the reduced name I have.

In [111]:
cleandata2 = cleandata1.copy()

In [140]:
name = cleandata2.iloc[3002440]['SOC_NAME'] #example of accessing a location
print("Before: "+name)
newname = reducedf.loc[(reducedf['SOC_NAME']==name)]
newname1 = newname.iloc[0]['Name2']
print("After: "+newname1)

software developers, applications
software developers


Test code to test algorithm

In [ ]:
errormap = defaultdict(list)

for index, row in cleandata2.iterrows():
    name = row['SOC_NAME']
    newname = reducedf.loc[(reducedf['SOC_NAME']==name)]
    try:
        newname1 = newname.iloc[0]['Name2']
        cleandata2.set_value([index],['SOC_NAME'],newname1)
    except:
        quicklist = ["Something went wrong", name, newname]
        errormap[index].append(quicklist) 

Here, we try to apply what we have reduced to the dataframe. We will see if we have made any progress, and what we can do next.

In [144]:
cleandata2['SOC_NAME'].value_counts().count()

1575

Debating what will happen if I remove all entries in the dataset with which a SOC_NAME that only occures once. How many entries will this remove? Can we ignore entries like this?

What about entries with an occurence count of 2? 3? Do these entries matter?

What do we do if they matter?